flashinfer/comm/trtllm_mnnvl_ar.py

```diff
@@ -25,102 +27,451 @@ def mpi_barrier():
     MPI.COMM_WORLD.Barrier()
 
 
+class MNNVLAllreduceFusionStrategy(Enum):
+    ONESHOT = 0
+    TWOSHOT = 1
+    AUTO = 99
+
+    @staticmethod
+    def select_strategy(
+        tp_size: int, num_tokens: int, hidden_dim: int, dtype: torch.dtype
+    ) -> "MNNVLAllreduceFusionStrategy":
+        elem_size = torch.tensor([], dtype=dtype).element_size()
+        if num_tokens * hidden_dim * tp_size * elem_size <= MNNVL_ONE_SHOT_THRESHOLD:
+            return MNNVLAllreduceFusionStrategy.ONESHOT
+        else:
+            return MNNVLAllreduceFusionStrategy.TWOSHOT
+
+
+# Empirical result calculated from num_tokens * hidden_dim * tp_size * elem_size
+MNNVL_ONE_SHOT_THRESHOLD = 64 * 1024 * 8 * 2
+
+
+class MNNVLAllreduceFusionWorkspace:
+    NUM_LAMPORT_BUFFERS = 3
+
+    def __init__(
+        self,
+        mapping: Mapping,
+        max_num_tokens: Optional[int] = None,
+        hidden_dim: Optional[int] = None,
+        dtype: Optional[torch.dtype] = None,
+        buffer_size_in_bytes: Optional[int] = None,
+        comm_backend: Optional[CommBackend] = None,
+    ):
+        """
+        Initialize the MNNVL Allreduce Fusion Workspace. comm_backend will be used for creating the workspace and synchronization. If not provided, MPIBackend will be used which will use COMM_WORLD for synchronization.
+
+        Args:
+            mapping: Mapping configuration containing rank info
+            buffer_size_in_bytes: The requested size in bytes for each lamport buffer. The actual allocation size may be larger due to alignment requirements. The actual usable size will be NUM_LAMPORT_BUFFERS * actual_buffer_size_per_lamport_buffer.
+        """
+
+        if buffer_size_in_bytes is None:
+            assert (
+                max_num_tokens is not None
+                and hidden_dim is not None
+                and dtype is not None
+            ), (
+                "max_num_tokens, hidden_dim, and dtype must be provided if buffer_size_in_bytes is not provided."
+            )
+
+            # If the user want to explictly use one-shot pass the threshold, which requires larger workspace size,
```

Comment:
Maybe we can add a check for this:
if (one_shot_size_bytes > MNNVL_ONE_SHOT_THRESHOLD) and strategy == MNNVLAllreduceFusionStrategy.ONESHOT:
raise ValueError("Please set the workspace when using oneshot beyond treshold).